Models usage docs revisions (#5738)
pydantic/pydantic: docs/usage/models.md

The primary means of defining objects in Pydantic is via models. Models are simply classes which inherit from
`BaseModel`.

You can think of models as similar to types in strictly typed languages, or as the requirements of a single endpoint
in an API.

Untrusted data can be passed to a model and, after parsing and validation, Pydantic guarantees that the fields
of the resultant model instance will conform to the field types defined on the model.

!!! note
    Pydantic is primarily a parsing library, **not a validation library**.
    Validation is a means to an end: building a model which conforms to the types and constraints provided.

>     In other words, Pydanti guarantees the types and constraints of the output model, not the input data.

Reviewer: ```suggestion
    In other words, Pydantic guarantees the types and constraints of the output model, not the input data.
```

    This might sound like an esoteric distinction, but it is not. If you're unsure what this means or
    how it might affect your usage you should read the section about [Data Conversion](#data-conversion) below.

>     Although validation is not the main purpose of Pydantic, you **can** use this library for custom [validation](validators.md).

Reviewer: This may be a good place to address the fact that we use the word `validator` in many places to refer to something that might more accurately be described as "transforming". @samuelcolvin not sure how you are thinking about this these days..

    name: str = 'Jane Doe'
```

In this example, `User` is a model with two fields:

* `id`, which is an integer and is required
* `name`, which is a string and is not required (it has a default value).

> The type of `name` is inferred from the
> default value, and so a type annotation is not required (however, please see the warning on
> [Field Ordering](#field-ordering) regarding field order when some fields do not have type annotations).

Reviewer: ```suggestion
```

I think let's just get rid of this comment now, as it no longer applies (in v2, all fields _must_ have annotations).

```py group="basic-model"
user = User(id='123')
```

In this example, `user` is an instance of `User`.
Initialisation of the object will perform all parsing and validation.
If no `ValidationError` is raised, you know the resulting model instance is valid.

```py group="basic-model"
assert user.id == 123
assert isinstance(user.id, int)

> # Note that '123' was cast to an int and its value is 123

Reviewer: ```suggestion
# Note that '123' was coerced to an int and its value is 123
```

```

> More details on casting can be found in [Data Conversion](#data-conversion).

Reviewer: ```suggestion
More details on pydantic's coercion logic can be found in [Data Conversion](#data-conversion).
```

This model is mutable so field values can be changed.

### Model methods and properties

The example above only shows the tip of the iceberg of what models can do.
Models possess the following methods and attributes:

> - `copy()`: returns a copy (by default, shallow copy) of the model. See

Reviewer: ```suggestion
- `model_copy()`: returns a copy (by default, shallow copy) of the model. See
```
Not sure if this is right, but there is a `model_copy` method and I assume that's what we want here?

(Also, I do like that you've included them in alphabetical order and would encourage that, I also think we should reorder them in the actual `main.py` file..)

For self-referencing models, see [postponed annotations](postponed_annotations.md#self-referencing-models).

## Arbitrary class instances

> (Formerly known as "ORM Mode", or building models from attributes.)

Reviewer: ```suggestion
(Formerly known as "ORM Mode"/`from_orm`)
```

> (Formerly known as "ORM Mode", or building models from attributes.)

Reviewer: ```suggestion
(Formerly known as "ORM Mode"/`from_orm`)
```

Pydantic models can be created from arbitrary class instances to support models that map to ORM objects.

To do this, use the `model_config` property on the model with `from_attributes` set to `True`. See
[Model Config](model_config.md) and [ConfigDict](../api/config.md) for more information.

> The example here uses [SQLAlchemy](https://www.sqlalchemy.org/), but the same approach should work for any ORM.

Reviewer: For what it's worth, I think we should focus less intently on ORM here — I would phrase it as "Pydantic models can also be created from arbitrary class instances by reading the instance attributes corresponding to the model field names. One common application of this functionality is integration with ORMs." or similar; right now it seems like this functionality is specific to ORMs but it's actually much more general, and I think with the recent name changes, imo, it's worth introducing it in that way.
